### PR TITLE
Replace bean method calls with injection

### PR DESCRIPTION
**Pull request by eleftherias**
This is so that our configuration classes do not rely on CGLIB to proxy bean methods.

Fixes gh-6818

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java [hunk 1]
```diff
@@ -103,12 +103,10 @@ public static InitializeAuthenticationProviderBeanManagerConfigurer initializeAu
 		return new InitializeAuthenticationProviderBeanManagerConfigurer(context);
 	}
 
-	public AuthenticationManager getAuthenticationManager() throws Exception {
+	public AuthenticationManager getAuthenticationManager(AuthenticationManagerBuilder authBuilder) throws Exception {
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java]: This is a major API change so it could only be allowed in Spring Security 6.0. Is this issue scheduled for 6.0? For more info around versioning see [Semantic Versioning](https://semver.org/spec/v2.0.0.html)

File: config/src/main/java/org/springframework/security/config/annotation/method/configuration/GlobalMethodSecurityConfiguration.java [hunk 2]
```diff
@@ -119,19 +115,19 @@
...
 	 * {@link MethodInterceptor}.
 	 * </p>
+	 * @param methodSecurityMetadataSource the default {@link MethodSecurityMetadataSource}.
 	 *
 	 * @return the {@link MethodInterceptor}.
-	 * @throws Exception
 	 */
 	@Bean
-	public MethodInterceptor methodSecurityInterceptor() throws Exception {
+	public MethodInterceptor methodSecurityInterceptor(MethodSecurityMetadataSource methodSecurityMetadataSource) {
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/method/configuration/GlobalMethodSecurityConfiguration.java]: Same reason as above re: Semantic Versioning

File: config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java [hunk 3]
```diff
@@ -64,15 +62,13 @@
  * @since 3.2
  */
 @Configuration
-public class WebSecurityConfiguration implements ImportAware, BeanClassLoaderAware {
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java]: Same reason as above re: Semantic Versioning

Comment (eleftherias) [reply config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java]: This change isn't neccessary, I did not consider the configurations that are extending this one. Will revert it back to implementing `BeanClassLoaderAware`.

File: config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java (hunk 1 above)
Comment (rwinch) [reply config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java]: I struggle with this a little bit because we want to guarantee that the `AuthenticationManagerBuilder` is actually a `Bean`. Perhaps we could change to do this:

```java
AuthenticationManagerBuilder authBuilder = this.applicationContext.getBean(AuthenticationManagerBuilder.class);
```

File: config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java [hunk 4]
```diff
@@ -123,16 +120,19 @@ public WebInvocationPrivilegeEvaluator privilegeEvaluator() throws Exception {
...
+	 * {@link AutowiredWebSecurityConfigurersIgnoreParents} containing the
 	 * {@code <SecurityConfigurer<FilterChainProxy, WebSecurityBuilder>} instances used to
 	 * create the web configuration
 	 * @throws Exception
 	 */
 	@Autowired(required = false)
 	public void setFilterChainProxySecurityConfigurer(
 			ObjectPostProcessor<Object> objectPostProcessor,
-			@Value("#{@autowiredWebSecurityConfigurersIgnoreParents.getWebSecurityConfigurers()}") List<SecurityConfigurer<Filter, WebSecurity>> webSecurityConfigurers)
+			AutowiredWebSecurityConfigurersIgnoreParents autowiredWebSecurityConfigurersIgnoreParents)
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java]: Can you explain why this change is necessary? A concern I have is that setting `List<SecurityConfigurer<Filter, WebSecurity>>` is agnostic to have we got it and only the annotation instructs how to look up the value. Another concern is that `AutowiredWebSecurityConfigurersIgnoreParents` is package private.

File: config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java [hunk 5]
```diff
@@ -71,8 +70,6 @@
 
 	private List<SecurityConfigurer<Filter, WebSecurity>> webSecurityConfigurers;
 
-	private ClassLoader beanClassLoader;
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java]: I see that we don't need beanClassLoader, but since it is unrelated can we please put it in another PR?

File: config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java [hunk 6]
```diff
@@ -43,7 +43,7 @@
  * @since 5.0
  */
 @Configuration
-class ServerHttpSecurityConfiguration implements WebFluxConfigurer {
+class ServerHttpSecurityConfiguration {
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java]: Can you explain why we need to expose this via a Bean vs the Configuration? Does setting proxy proxyBeanMethods mean the `ServerHttpSecurityConfiguration` is no longer registered as a Bean?

File: config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java [hunk 7]
```diff
@@ -113,6 +119,10 @@ public CurrentSecurityContextArgumentResolver reactiveCurrentSecurityContextArgu
 
 	@Bean(HTTPSECURITY_BEAN_NAME)
 	@Scope("prototype")
+	public ServerHttpSecurity serverHttpSecurity() {
```
[the remark this answers is not in the document]
Comment (eleftherias) [reply config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java]: There is a test that uses the `httpSecurity()` method. 
https://github.com/spring-projects/spring-security/blob/047bd16b5156de0bb595ff6b324a63cef92b1fe1/config/src/test/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfigurationBuilder.java#L31
If we separate the bean method from the method that creates the object then we avoid calling the bean method in the test.
Unless we aren't concerned about the test using the bean method because it's creating a new instance of `ServerHttpSecurityConfiguration`?

File: config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java (hunk 6 above)
Comment (eleftherias) [reply config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java]: `ServerHttpSecurityConfiguration` will still be registered as a Bean. 
The reason for this change is to avoid calling `authenticationPrincipalArgumentResolver()` from the `configureArgumentResolvers` override. 
I noticed spring boot had a similar situation when disabling proxying and this was their approach 
https://github.com/spring-projects/spring-boot/blob/dfb2e8a278e7211c658104ae45a19b68aec57a21/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jDataAutoConfiguration.java#L136

File: config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rwinch) [reply config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java]: When it just performs new it won't create a Bean since nothing is processing the annotation.

File: config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java (hunk 6 above)
Comment (rwinch) [reply config/src/main/java/org/springframework/security/config/annotation/web/reactive/ServerHttpSecurityConfiguration.java]: Ok thanks for the clarification. I think that this makes sense.

File: config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java (hunk 4 above)
Comment (eleftherias) [reply config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java]: This change is not neccessary. I have reverted it.

File: config/src/main/java/org/springframework/security/config/annotation/method/configuration/GlobalMethodSecurityConfiguration.java (hunk 2 above)
Comment (rwinch) [reply config/src/main/java/org/springframework/security/config/annotation/method/configuration/GlobalMethodSecurityConfiguration.java]: I don't think this is avoidable. Given the high priority of this and the low likely hood of causing users any problems, I think this breaking change is ok.

**Comment by eleftherias** [comment]
@rwinch Thanks for pointing that out. I limited my search to `@Configuration`, didn't think about classes that could be using other annotations!

File: config/src/test/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfigurationTests.java [hunk 8]
```diff
@@ -172,51 +171,49 @@ public AuthenticationManager authenticationManager() {
 			return this.authenticationManager;
 		}
 	}
-		//
```
Comment (rwinch) [inline config/src/test/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfigurationTests.java]: There seem to be some formatting changes in the code within this commit. While overall, they are a good thing I think we should move it to another PR or at minimum a separate commit. The reason is it helps us know the intent of the changes when being reviewed or when looking through commit history if we are trying to figure out a bug at some point.

**Comment by rwinch** [comment]
Thanks @eleftherias! This is now merged into master
